Runtime/Scripts/DebugUtils/RGGizmosWrapper.cs

```diff
@@ -0,0 +1,383 @@
+            {
+                if (sphereParams.Item4)
+                {
+                    Gizmos.color = sphereParams.Item2;
+                    Gizmos.DrawWireSphere(sphereParams.Item1, sphereParams.Item3);
+                }
+                else
+                {
+                    Gizmos.color = sphereParams.Item2;
+                    Gizmos.DrawSphere(sphereParams.Item1, sphereParams.Item3);
+                }
+            }
+            foreach (var sphereParams in _spheresAtEntity.Values)
+            {
+                var originInstance = RGFindUtils.Instance.FindOneByInstanceId<RGEntity>(sphereParams.Item1);
+                if (originInstance == null) continue;
+                if (sphereParams.Item4)
+                {
+                    Gizmos.color = sphereParams.Item2;
+                    Gizmos.DrawWireSphere(originInstance.transform.position, sphereParams.Item3);
+                }
+                else
+                {
+                    Gizmos.color = sphereParams.Item2;
+                    Gizmos.DrawSphere(originInstance.transform.position, sphereParams.Item3);
+                }
+            }
+            
+            // Delete any billboards that are no longer being drawn
+            var billboardsToRemove = _drawnBillboards.Where(b => !_billboardsToDraw.ContainsKey(b.Key));
+            foreach (var billboard in billboardsToRemove)
+            {
+                Object.Destroy(billboard.Value);
+                _drawnBillboards.Remove(billboard.Key);
+            }
+            
+            // Now update or create any billboards that are being drawn
+            foreach (var billboardParams in _billboardsToDraw)
+            {
+                try
+                {
+                    // If the billboard does not exist, create it
+                    if (!_drawnBillboards.ContainsKey(billboardParams.Key))
+                    {
+                        var entityGameObject = RGFindUtils.Instance.FindOneByInstanceId<RGEntity>(billboardParams.Key);
+                        if (entityGameObject == null) continue;
+                        var billboardObject = Object.Instantiate(_billboardAsset, entityGameObject.transform.position,
+                            Quaternion.identity);
+                        billboardObject.transform.SetParent(entityGameObject.transform);
+                        _drawnBillboards[billboardParams.Key] = billboardObject;
+                    }
+                
+                    // Then set the parameters
+                    var billboardText = _drawnBillboards[billboardParams.Key].GetComponent<BillboardText>();
+                    billboardText.SetText(billboardParams.Value.Item1);
+                
+                    // If an offset is given, use that for placing it above the agent
+                    billboardText.SetYOffset(billboardParams.Value.Item2);
+                }
+                catch (Exception e)
```

Comment:
Yeah good point... got lazy here and you caught me 😬. Basically I am trying to handle the situation when an object is deleted from the scene, and so `_drawnBillboards[billboardParams.Key]` may be destroyed. I'll test this scenario and see what the exception is!

Reply:
Discovered the exception and updated! During this process, I also found a concurrency issue... which should be very uncommon with how this code should work, but still happened. Added ConcurrentDictionary over Dictionary.

Reply:
Yeah, if we know there are concurrent invocations at play, using a `ConcurrentDictionary` is probably best. That will change some of the patterns though since you'll need to avoid races.

Reply:
> Basically I am trying to handle the situation when an object is deleted from the scene, and so _drawnBillboards[billboardParams.Key] may be destroyed. I'll test this scenario and see what the exception is!

It'd be ideal if we could check if the billboard has been destroyed _before_ attempting to work with it. Something like:

```csharp
var billboard = _drawnBillboards[billboardParams.Key];

if(!billboard.IsDestroyed) {
    // ... work with it
}
```

That means that in _most_ cases, the exception is never thrown. Exceptions are _quite_ expensive in .NET (though to be fair, that cost probably isn't really noticable here) and it's best to avoid them entirely as much as possible. Of course, if there's concurrent code at play here, that pattern I showed above wouldn't be perfect and we'd still want the `try..catch` inside the `if(!billboard.IsDestroyed)` to cover unexpected cases, but we'd be able to avoid the Exception in _most_ cases and that's ideal.

Reply:
Thanks for the suggestion! Added this to my code. Btw, once you are satisfied with changes as it relates to a comment, make sure that you mark them as resolved - we have the comment writer, not the PR owner, mark conversations as resolved so they can verify that the PR owner made the requested changes properly.